src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractLanguageService`2.VsCodeWindowManager.cs

```diff
@@ -22,21 +29,71 @@ internal class VsCodeWindowManager : IVsCodeWindowManager, IVsCodeWindowEvents
+            private readonly IAsynchronousOperationListener _asynchronousOperationListener;
 
-            private INavigationBarController _navigationBarController;
-            private IVsDropdownBarClient _dropdownBarClient;
+            private INavigationBarController? _navigationBarController;
+            private IVsDropdownBarClient? _dropdownBarClient;
+            private IOptionService? _optionService;
+            private WorkspaceRegistration? _workspaceRegistration;
 
             public VsCodeWindowManager(TLanguageService languageService, IVsCodeWindow codeWindow)
             {
                 _languageService = languageService;
                 _codeWindow = codeWindow;
 
-                var workspace = languageService.Package.ComponentModel.GetService<VisualStudioWorkspace>();
-                _optionService = workspace.Services.GetService<IOptionService>();
+                _threadingContext = languageService.Package.ComponentModel.GetService<IThreadingContext>();
+
+                var listenerProvider = languageService.Package.ComponentModel.GetService<IAsynchronousOperationListenerProvider>();
+                _asynchronousOperationListener = listenerProvider.GetListener(FeatureAttribute.NavigationBar);
 
                 _sink = ComEventSink.Advise<IVsCodeWindowEvents>(codeWindow, this);
-                _optionService.OptionChanged += OnOptionChanged;
+            }
+
+            private void OnWorkspaceRegistrationChanged(object sender, System.EventArgs e)
+            {
+                var token = _asynchronousOperationListener.BeginAsyncOperation(nameof(OnWorkspaceRegistrationChanged));
+
+                // Fire and forget to update the navbar based on the workspace registration
+                // to avoid blocking the caller and possible deadlocks workspace registration changed events under lock.
+                UpdateWorkspace().CompletesAsyncOperation(token).Forget();
+            }
+
+            private async Task UpdateWorkspace()
+            {
+                // This event may not be triggered on the main thread, but adding and removing the navbar
+                // must be done from the main thread.
+                await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                // If the workspace registration is missing, addornments have been removed.
+                if (_workspaceRegistration == null)
+                {
+                    return;
+                }
+
+                // There's a new workspace, so make sure we unsubscribe from the old workspace option changes and subscribe to new.
+                UpdateOptionChangedSource(_workspaceRegistration.Workspace);
+
+                _navigationBarController?.SetWorkspace(_workspaceRegistration.Workspace);
+
+                // Trigger a check to see if the dropdown should be added / removed now that the buffer is in a different workspace.
+                AddOrRemoveDropdown();
```

Comment:
Inline this into UpdateOptionChangedSource, since all the callers probably have to call it?

Reply:
(this might be tricky if you have to move the _navigationBarController?.SetWorkspace() line up earlier..)

Reply:
I'm going to skip this one - AddOrRemoveDropdown is also called in response to OnOptionChanged events which do not need to update the option service source.